notifications/manager.go

```diff
@@ -0,0 +1,214 @@
+package notifications
+
+import (
+	"context"
+	"crypto/rand"
+	"encoding/hex"
+	"time"
+
+	"github.com/lightninglabs/loop/swapserverrpc"
+)
+
+type NotificationType int
+
+const (
+	NotificationTypeUnknown NotificationType = iota
+	NotificationTypeReservation
+)
+
+// Config contains all the services that the notification manager needs to
+// operate.
+type Config struct {
+	// Client is the client used to communicate with the swap server.
+	Client swapserverrpc.SwapServerClient
+
+	// FetchL402 is the function used to fetch the l402 token.
+	FetchL402 func(context.Context) error
+}
+
+// Manager is a manager for notifications that the swap server sends to the
+// client.
+type Manager struct {
+	cfg *Config
+
+	hasL402 bool
+
+	runCtx context.Context
+
+	subscribers map[NotificationType]map[string]func(*swapserverrpc.SubscribeNotificationsResponse) //nolint:lll
+}
+
+// NewManager creates a new notification manager.
+func NewManager(cfg *Config) *Manager {
+	return &Manager{
+		cfg:         cfg,
+		subscribers: make(map[NotificationType]map[string]func(*swapserverrpc.SubscribeNotificationsResponse)), //nolint:lll
+	}
+}
+
+// SubscribeReservations subscribes to the reservation notifications.
+func (n *Manager) SubscribeReservations(
+	callback func(*swapserverrpc.ServerReservationNotification)) string {
+
+	fn := func(notif *swapserverrpc.SubscribeNotificationsResponse) {
+		callback(notif.GetReservationNotification())
+	}
+	return n.subscribe(NotificationTypeReservation, fn)
+}
+
+// subscribe subscribes to a notification type and returns a unique ID that can
+// be used to unsubscribe.
+func (n *Manager) subscribe(notificationType NotificationType,
+	callback func(*swapserverrpc.SubscribeNotificationsResponse)) string {
+	if n.subscribers[notificationType] == nil {
+		n.subscribers[notificationType] = make(map[string]func(*swapserverrpc.SubscribeNotificationsResponse)) //nolint:lll
+	}
+
+	// Generate a unique ID for this subscription
+	id := generateUniqueID()
+
+	n.subscribers[notificationType][id] = callback
+
+	return id
+}
+
+// Unsubscribe removes a subscription based on its ID.
+func (n *Manager) Unsubscribe(id string) {
+	for _, subscribers := range n.subscribers {
+		delete(subscribers, id)
+	}
+}
+
+// Run starts the notification manager.
+func (n *Manager) Run(ctx context.Context, readyChan chan struct{},
+) error {
+
+	n.runCtx = ctx
+
+	err := n.RegisterNotifications()
+	if err != nil {
+		return err
+	}
+
+	// Notify the caller that we're ready, this means we have successfully
+	// fetched a l402 token and have subscribed to the notifications at least
+	// once.
+	close(readyChan)
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		}
+	}
```

Comment:
Can we just run the event loop here instead of creating a goroutine?

This will also fix the lint warning:
```
S1000: should use for range instead of for { select {} } (gosimple)
```

Reply:
I propose to re-organize the code to have only one event loop and avoid new goroutines launch.

In case a reconnect is needed, the code in that loop can do the actions needed to reconnect and reassign some variables and then proceed with receiving notifications in new iteration.

I think it will simplify the code and make it more reader friendly. It would be easier to follow the flow of code.